src/compileout.c

```diff
@@ -0,0 +1,442 @@
+*/
+
+#include <stdlib.h>
+#include <assert.h>
+
+#include "julia.h"
+#include "julia_internal.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+JL_DLLEXPORT int jl_generating_output(void)
+{
+    return jl_options.outputo || jl_options.outputbc || jl_options.outputji;
+}
+
+void jl_precompile(int all);
+
+void jl_write_compiler_output(void)
+{
+    if (!jl_generating_output())
+        return;
+
+    if (!jl_options.incremental)
+        jl_precompile(jl_options.compile_enabled == JL_OPTIONS_COMPILE_ALL);
+
+    if (!jl_module_init_order) {
+        jl_printf(JL_STDERR, "WARNING: --output requested, but no modules defined during run\n");
+        return;
+    }
+
+    jl_array_t *worklist = jl_module_init_order;
+    JL_GC_PUSH1(&worklist);
+    jl_module_init_order = jl_alloc_vec_any(0);
+    int i, l = jl_array_len(worklist);
+    for (i = 0; i < l; i++) {
+        jl_value_t *m = jl_arrayref(worklist, i);
+        if (jl_get_global((jl_module_t*)m, jl_symbol("__init__"))) {
+            jl_array_ptr_1d_push(jl_module_init_order, m);
+        }
+    }
+
+    if (jl_options.incremental) {
+        if (jl_options.outputji)
+            if (jl_save_incremental(jl_options.outputji, worklist))
+                jl_exit(1);
+        if (jl_options.outputbc)
+            jl_printf(JL_STDERR, "WARNING: incremental output to a .bc file is not implemented\n");
+        if (jl_options.outputo)
+            jl_printf(JL_STDERR, "WARNING: incremental output to a .o file is not implemented\n");
+    }
+    else {
+        ios_t *s = NULL;
+        if (jl_options.outputo || jl_options.outputbc)
+            s = jl_create_system_image();
+
+        if (jl_options.outputji) {
+            if (s == NULL) {
+                jl_save_system_image(jl_options.outputji);
+            }
+            else {
+                ios_t f;
+                if (ios_file(&f, jl_options.outputji, 1, 1, 1, 1) == NULL)
+                    jl_errorf("cannot open system image file \"%s\" for writing", jl_options.outputji);
+                ios_write(&f, (const char*)s->buf, (size_t)s->size);
+                ios_close(&f);
+            }
+        }
+
+        if (jl_options.outputo || jl_options.outputbc)
+            jl_dump_native(jl_options.outputbc,
+                           jl_options.outputo,
+                           (const char*)s->buf, (size_t)s->size);
+    }
+    JL_GC_POP();
+}
+
+int jl_is_rettype_inferred(jl_method_instance_t *li);
+
+JL_DLLEXPORT int jl_compile_hint(jl_tupletype_t *types)
+{
+    size_t world = jl_world_counter;
+    jl_method_instance_t *li = jl_get_specialization1(types, world);
+    if (li == NULL)
+        return 0;
+    jl_code_info_t *src = NULL;
+    if (!jl_is_rettype_inferred(li))
+        src = jl_type_infer(&li, world, 0);
+    if (li->jlcall_api != 2)
+        jl_compile_linfo(&li, src, world, &jl_default_cgparams);
+    return 1;
+}
+
+#if 0  // TODO restore this for jb/subtype
```

Comment:
probably should be restored, in order to make sure this code movement isn't breaking anything here